### PR TITLE
Jakim API migration

### PR DESCRIPTION
**Pull request by iqfareez**
Yes, again. For the sake of long-term reliability

### CODE REVIEW

**Comment by iqfareez** [comment]
Work in progress :construction:

image

**Comment by iqfareez** [comment]
> Work in progress 🚧
> 
> image
> 
> (Updating location database)

Xjadi

**Comment by iqfareez** [comment]
Changing how location are stored in device. Previously I save the index, egg `0` for JHR01, `49` for SGR01 and so on. This time, I want to save the actual location code, eg: `JHR01`, `JHR02` etc. As I said, it is a long-term migration, just in case JAKIM will add a new zone, it didn't take much to implement it.

File: lib/notificationUtil/notification_scheduler.dart [hunk 1]
```diff
@@ -143,21 +144,21 @@ class MyNotifScheduler {
...
+          DateTime.fromMillisecondsSinceEpoch(dayTime.times![2]);
       DateTime asarDateTime =
-          DateTime.fromMillisecondsSinceEpoch(dayTime[3] * 1000);
+          DateTime.fromMillisecondsSinceEpoch(dayTime.times![3]);
       DateTime maghribDateTime =
-          DateTime.fromMillisecondsSinceEpoch(dayTime[4] * 1000);
+          DateTime.fromMillisecondsSinceEpoch(dayTime.times![4]);
       DateTime isyakDateTime =
-          DateTime.fromMillisecondsSinceEpoch(dayTime[5] * 1000);
+          DateTime.fromMillisecondsSinceEpoch(dayTime.times![5]);
```
Comment (iqfareez) [inline lib/notificationUtil/notification_scheduler.dart:161]: X perasan part ni sblum ni patut la notification salah2 haaaha

**Comment by iqfareez** [comment]
Alhamdulillah, released update 2.2.2

I'll take more about this whole purpose for API migration here, later.

**Comment by iqfareez** [comment]
## Reason behind this migration (to JAKIM API)

_Perhaps useful to some people or even me myself in the future :wink:_

Let's start with the problems. For this app, I've been using 2-3 different APIs. There are always things that I like and things I don't. Common weaknesses in third-party APIs (Azanpro, WaktuSolatApp, and mpti906) that I found is that the prayer times is **outdated from JAKIM prayer times**. JAKIM always makes amendment and correction to their prayer times over time, and most third-party APIs that take some times (a month or more) to process those changes. The **precision of prayer times** is vital for a Muslim, and it makes me uncomfortable if this MPT app shows the wrong prayer times. The advantages of these third-party APIs (except WaktuSolatApp API) are having good documentation and reference.

So, that leaves us with JAKIM API, which we can guarantee is the most reliable source of truth for Malaysia prayer times. Yes, but the problem (or weakness) is their API is not public by default; I never saw the docs to use the API. People found those API by inspecting the network traffic in DevTools. So, there is a risk that JAKIM might change their endpoint without notice, and hence breaking the app (oh, I might just overthink this). 

_I have more to say here, but I don't know how to say hahaha. And sorry but I feel nobody will understand incl. myself in the future hahahaha_

Oh, did I mention that JAKIM API also most of the time [went down](https://iqfareez.github.io/jakim-esolat-status)? They have excellent reliability on the database but are poor at managing server loads etc. 

So, in conclusion, I chose JAKIM API because of its data reliability for a **long run**, inshallah, and to overcome the aforementioned problem, I made a [backup API](https://mpt-backup-api.herokuapp.com/), which is just a 'clone' to the JAKIM API just in case their server went offline, this backup API will be used.
